compiler/rustc_middle/src/ty/context/tls.rs

```diff
@@ -0,0 +1,231 @@
+fn erase(context: &ImplicitCtxt<'_, '_>) -> *const () {
+    context as *const _ as *const ()
+}
+
+unsafe fn downcast<'a, 'tcx>(context: *const ()) -> &'a ImplicitCtxt<'a, 'tcx> {
+    &*(context as *const ImplicitCtxt<'a, 'tcx>)
+}
+
+/// Sets `context` as the new current `ImplicitCtxt` for the duration of the function `f`.
+#[inline]
+pub fn enter_context<'a, 'tcx, F, R>(context: &ImplicitCtxt<'a, 'tcx>, f: F) -> R
+where
+    F: FnOnce(&ImplicitCtxt<'a, 'tcx>) -> R,
+{
+    tlv::with_tlv(erase(context), || f(&context))
+}
+
+/// Allows access to the current `ImplicitCtxt` in a closure if one is available.
+#[inline]
+pub fn with_context_opt<F, R>(f: F) -> R
+where
+    F: for<'a, 'tcx> FnOnce(Option<&ImplicitCtxt<'a, 'tcx>>) -> R,
+{
+    let context = tlv::get_tlv();
+    if context.is_null() {
+        f(None)
+    } else {
+        // We could get an `ImplicitCtxt` pointer from another thread.
+        // Ensure that `ImplicitCtxt` is `Sync`.
+        sync::assert_sync::<ImplicitCtxt<'_, '_>>();
+
+        unsafe { f(Some(downcast(context))) }
+    }
+}
+
+/// Replaces the current context with `context`. Returns a drop guard that resets the context when it's dropped.
+///
+/// You should use the closure API of [`enter_context`] instead.
+///
+/// # Safety
+/// The caller has to ensure that the drop guard is actually run before the lifetime of `context` ends.
+#[inline]
+pub unsafe fn set_scoped_context<'a, 'tcx>(context: &'a ImplicitCtxt<'a, 'tcx>) -> impl Sized {
+    let old = tlv::get_tlv();
+    let reset = rustc_data_structures::OnDrop(move || tlv::set_tlv(old));
+
+    tlv::set_tlv(erase(context));
+
+    reset
+}
+
+/// Unsafely gets an unbounded reference to the current context.
+///
+/// You should use the closure API of [`with_context`] instead.
+///
+/// # Safety
+/// The caller has to ensure that the context is not used after it is freed.
+#[inline]
+pub unsafe fn get_context<'a, 'tcx>() -> &'a ImplicitCtxt<'a, 'tcx> {
+    let context = tlv::get_tlv();
+    if context.is_null() {
+        panic!("no ImplicitCtxt stored in tls");
+    } else {
+        // We could get an `ImplicitCtxt` pointer from another thread.
+        // Ensure that `ImplicitCtxt` is `Sync`.
+        sync::assert_sync::<ImplicitCtxt<'_, '_>>();
+
+        unsafe { downcast(context) }
+    }
+}
+
+/// Allows access to the current `ImplicitCtxt`.
+/// Panics if there is no `ImplicitCtxt` available.
+#[inline]
+pub fn with_context<F, R>(f: F) -> R
+where
+    F: for<'a, 'tcx> FnOnce(&ImplicitCtxt<'a, 'tcx>) -> R,
+{
+    with_context_opt(|opt_context| f(opt_context.expect("no ImplicitCtxt stored in tls")))
+}
+
+/// Allows access to the current `ImplicitCtxt` whose tcx field is the same as the tcx argument
+/// passed in. This means the closure is given an `ImplicitCtxt` with the same `'tcx` lifetime
+/// as the `TyCtxt` passed in.
+/// This will panic if you pass it a `TyCtxt` which is different from the current
+/// `ImplicitCtxt`'s `tcx` field.
+#[inline]
```

Comment:
Other idea: could we avoid unsafe code by using `inline(always)` on the 3 functions that take a closure. MIR inlining would reduce the current API to the unsafe one automatically.